mava/wrappers/gym.py

```diff
@@ -0,0 +1,312 @@
+        else:
+            reward = np.array(reward)
+
+        truncated = [truncated] * self.num_agents
+        terminated = [terminated] * self.num_agents
+
+        return agents_view, reward, terminated, truncated, info
+
+    def get_actions_mask(self, info: Dict) -> NDArray:
+        if "action_mask" in info:
+            return np.array(info["action_mask"])
+        return np.ones((self.num_agents, self.num_actions), dtype=np.float32)
+
+
+class GymRecordEpisodeMetrics(gym.Wrapper):
+    """Record the episode returns and lengths."""
+
+    def __init__(self, env: gym.Env):
+        super().__init__(env)
+        self._env = env
+        self.running_count_episode_return = 0.0
+        self.running_count_episode_length = 0.0
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[np.ndarray, Dict]:
+
+        # Reset the env
+        agents_view, info = self._env.reset(seed, options)
+
+        # Create the metrics dict
+        metrics = {
+            "episode_return": self.running_count_episode_return,
+            "episode_length": self.running_count_episode_length,
+            "is_terminal_step": True,
+        }
+
+        # Reset the metrics
+        self.running_count_episode_return = 0.0
+        self.running_count_episode_length = 0
+
+        if "won_episode" in info:
+            metrics["won_episode"] = info["won_episode"]
+
+        info["metrics"] = metrics
+
+        return agents_view, info
+
+    def step(self, actions: NDArray) -> Tuple:
+
+        # Step the env
+        agents_view, reward, terminated, truncated, info = self._env.step(actions)
+
+        self.running_count_episode_return += float(np.mean(reward))
+        self.running_count_episode_length += 1
+
+        metrics = {
+            "episode_return": self.running_count_episode_return,
+            "episode_length": self.running_count_episode_length,
+            "is_terminal_step": False,
+        }
+        if "won_episode" in info:
+            metrics["won_episode"] = info["won_episode"]
+
+        info["metrics"] = metrics
+
+        return agents_view, reward, terminated, truncated, info
+
+
+class GymAgentIDWrapper(gym.Wrapper):
+    """Add one hot agent IDs to observation."""
+
+    def __init__(self, env: gym.Env):
+        super().__init__(env)
+
+        self.agent_ids = np.eye(self.env.num_agents)
+        observation_space = self.env.observation_space[0]
+        _obs_low, _obs_high, _obs_dtype, _obs_shape = (
+            observation_space.low[0],
+            observation_space.high[0],
+            observation_space.dtype,
+            observation_space.shape,
+        )
+        _new_obs_shape = (_obs_shape[0] + self.env.num_agents,)
+        _observation_boxs = [
+            spaces.Box(low=_obs_low, high=_obs_high, shape=_new_obs_shape, dtype=_obs_dtype)
+        ] * self.env.num_agents
+        self.observation_space = spaces.Tuple(_observation_boxs)
```

Comment:
This will currently only work if the inner spaces are boxes. Can we make this more general?